server/src/main/java/org/elasticsearch/index/search/NestedHelper.java

```diff
@@ -53,14 +53,16 @@ public boolean mightMatchNestedDocs(Query query) {
             // We only handle term(s) queries and range queries, which should already
             // cover a high majority of use-cases
             return mightMatchNestedDocs(((TermQuery) query).getTerm().field());
-        } else if (query instanceof TermInSetQuery) {
-            PrefixCodedTerms terms = ((TermInSetQuery) query).getTermData();
-            if (terms.size() > 0) {
-                PrefixCodedTerms.TermIterator it = terms.iterator();
-                it.next();
-                return mightMatchNestedDocs(it.field());
-            } else {
-                return false;
+        } else if (query instanceof TermInSetQuery tis) {
+            try {
+                if (tis.getTermsCount() > 0) {
+                    return mightMatchNestedDocs(tis.getField());
+                } else {
+                    return false;
+                }
+            } catch (IOException e) {
+                // this handling isn't needed any more once we move to Lucene 10
+                throw new AssertionError("We are not doing IO here, this should never happen.");
```

Comment:
AssertionError will crash the node right? I do not like throwing this in a production code path. 

We can add an `assert false` here instead as that will only trip when assertions are enabled. 

As for what to do if we do get an IOException from `getTermsCount()` maybe rethrow as an `UncheckedIOException`

Reply:
agreed, either assert false or throw unchecked IO

Reply:
We cannot get an IOException, its never thrown and the Exception will be removed in Lucene, see https://github.com/apache/lucene/pull/13701
As for the AssertionException: this is quite common elsewhere when e.g. iterating BytesRefIterator [here in MessageDigest](https://github.com/elastic/elasticsearch/blob/main/server/src/main/java/org/elasticsearch/common/hash/MessageDigests.java#L144) or [here in Netty4Utils](https://github.com/elastic/elasticsearch/blob/main/modules/transport-netty4/src/main/java/org/elasticsearch/transport/netty4/Netty4Utils.java#L114).

That said, I don't care, will change it to unchecked IO fwiw

Reply:
> As for the AssertionException: this is quite common elsewhere when e.g. iterating BytesRefIterator

I disagree with those usages as well. Anything that has a "This should never happen" in the line makes me really uncomfortable. Its like asking for things to break :D. 

If we want to have assertions, we should use them. Then either ignore the thrown exception or throw a real exception with a real description encapsulating the actually thrown exception.